sherpa/utils/parallel.py

```diff
@@ -0,0 +1,462 @@
+    on a separate core and return the results (unless the machine only
+    has a single core or the parallel.numcores setting is set to 1).
+
+    >>> import numpy as np
+    >>> args = [np.arange(5), np.arange(3), np.arange(7)]
+    >>> parallel_map(np.sum, args)
+    [10, 3, 21]
+
+    Use two jobs to evaluate the results: one job will sum up two arrays
+    while the other will only sum one array since there are 3 jobs to
+    run.
+
+    >>> parallel_map(np.sum, args, numcores=2)
+    [10, 3, 21]
+
+    An example of sending in multiple arguments to a function (``comp``)
+    via a dictionary (although in this case there is only one task to
+    execute):
+
+    >>> parallel_map(comp, [{'idx1': 23, 'idx2': 47}])
+
+    Here the ``tcomp`` function accepts a single parameter which it
+    can deconstruct to extract the two values it needs:
+
+    >>> parallel_map(tcomp, [(23, 47), (2, 20), (5, 10)])
+
+    """
+    if not callable(function):
+        raise TypeError(f"input function '{repr(function)}' is not callable")
+
+    if not np.iterable(sequence):
+        raise TypeError(f"input '{repr(sequence)}' is not iterable")
+
+    size = len(sequence)
+
+    if not _multi or size == 1 or (numcores is not None and numcores < 2):
+        return list(map(function, sequence))
+
+    if numcores is None:
+        numcores = _ncpus
+
+    # Returns a started SyncManager object which can be used for sharing
+    # objects between processes. The returned manager object corresponds
+    # to a spawned child process and has methods which will create shared
+    # objects and return corresponding proxies.
+    manager = multiprocessing.Manager()
+
+    # Create FIFO queue and lock shared objects and return proxies to them.
+    # The managers handles a server process that manages shared objects that
+    # each slave process has access to.  Bottom line -- thread-safe.
+    out_q = manager.Queue()
+    err_q = manager.Queue()
+    # lock = manager.Lock() - currently unused
+
+    # if sequence is less than numcores, only use len sequence number of
+    # processes
+    if size < numcores:
+        numcores = size
+
+    # group sequence into numcores-worth of chunks
+    sequence = split_array(sequence, numcores)
+
+    procs = [multiprocessing.Process(target=worker,
+                                     args=(function, ii, chunk, out_q, err_q))
+             for ii, chunk in enumerate(sequence)]
+
+    return run_tasks(procs, err_q, out_q)
+
+
+# TODO: this routine needs a review
+def parallel_map_funcs(funcs, datasets, numcores=None):
+    """Run a sequence of function on a sequence of inputs in parallel.
+
+    Sherpa's parallel_map runs a single function to an iterable set of
+    sequence.  parallel_map_funcs is generalized parallelized version
+    of sherpa's parallel_map function since each element of the ordered
+    iterable funcs shall operate on the each element of the datasets.
```

Comment:
Below, it says that the length of the fundcs and datasets needs to be equal. That only makes sense if each func is applied to one-and-only-one if the datasets. In contrast, this section claims that each func operates on each dataset, i.e. that for n funcs and m datasets, I get n*m answers.
Only one can be right.

Reply:
Does this mean:
```suggestion
    of sherpa's parallel_map function. In `parallel_map_funcs` a different
    function can be associated with each dataset. For each dataset, the
    corresponding function is run for each element in that dataset.
```

Reply:
See https://github.com/sherpa/sherpa/pull/1776#discussion_r1209483512